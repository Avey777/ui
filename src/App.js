import React, {PropTypes} from 'react';
import {connect} from 'react-redux';
import SideNavContainer from 'src/side_nav';
import {
  publishNotification as publishNotificationAction,
  dismissNotification as dismissNotificationAction,
} from 'src/shared/actions/notifications';

const App = React.createClass({
  propTypes: {
    children: PropTypes.node.isRequired,
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }),
    params: PropTypes.shape({
      sourceID: PropTypes.string.isRequired,
    }).isRequired,
    publishNotification: PropTypes.func.isRequired,
    dismissNotification: PropTypes.func.isRequired,
    notifications: PropTypes.shape({
      success: PropTypes.string,
      error: PropTypes.string,
      warning: PropTypes.string,
    }),
  },

  handleNotification({type, text}) {
    this.props.publishNotification(type, text);
  },

  handleDismissNotification(type) {
    this.props.dismissNotification(type);
  },

  render() {
    const {sourceID} = this.props.params;

    return (
      <div className="enterprise-wrapper--flex">
        <SideNavContainer sourceID={sourceID} addFlashMessage={this.handleNotification} currentLocation={this.props.location.pathname} />
        <div className="page-wrapper">
          {this.renderNotifications()}
          {this.props.children && React.cloneElement(this.props.children, {
            addFlashMessage: this.handleNotification,
          })}
        </div>
      </div>
    );
  },

  renderNotifications() {
    const {success, error} = this.props.notifications;
    if (!success && !error) {
      return null;
    }
    if (success) {
      return (
        <div className="flash-messages">
          <div className="alert alert-success" role="alert">
            {success}{this.renderDismiss()}
          </div>
        </div>
      );
    }
    if (error) {
      return (
        <div className="flash-messages">
          <div className="alert alert-danger" role="alert">
            {error}{this.renderDismiss()}
          </div>
        </div>
      );
    }
  },

  renderDismiss() {
    return (
      <button className="close" data-dismiss="alert" aria-label="Close" onClick={this.handleDismissNotification}>
        <span className="icon remove"></span>
      </button>
    );
  },
});

function mapStateToProps(state) {
  return {
    notifications: state.notifications,
  };
}

export default connect(mapStateToProps, {
  publishNotification: publishNotificationAction,
  dismissNotification: dismissNotificationAction,
})(App);
