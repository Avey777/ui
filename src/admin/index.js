import UsersPage from './containers/UsersPage';
export {UsersPage};
