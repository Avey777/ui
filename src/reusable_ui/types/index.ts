export enum ComponentColor {
  Default = 'default',
  Primary = 'primary',
  Success = 'success',
  Warning = 'warning',
  Danger = 'danger',
  Alert = 'alert',
}

export enum ComponentSize {
  ExtraSmall = 'xs',
  Small = 'sm',
  Medium = 'md',
  Large = 'lg',
}

export enum ButtonShape {
  Default = 'none',
  Square = 'square',
  StretchToFit = 'stretch',
}
