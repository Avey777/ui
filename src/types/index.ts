import {AuthLinks, Organization, Role, User, Me} from './auth'
import {AlertRule, Kapacitor} from './kapacitor'
import {Query, QueryConfig} from './query'
import {Source} from './sources'
import {DropdownAction, DropdownItem} from './shared'

export {
  Me,
  AuthLinks,
  Role,
  User,
  Organization,
  AlertRule,
  Kapacitor,
  Query,
  QueryConfig,
  Source,
  DropdownAction,
  DropdownItem,
}
