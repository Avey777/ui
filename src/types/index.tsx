import {AuthLinks, Role, User, Organization} from './auth'
import {AlertRule, Kapacitor} from "./kapacitor"
import {Query, QueryConfig} from "./query"
import {Source} from "./sources"

export {AuthLinks, Role, User, Organization, AlertRule, Kapacitor, Query, QueryConfig, Source, }
