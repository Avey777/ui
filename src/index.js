import React from 'react'
import {render} from 'react-dom'
import {Provider} from 'react-redux'
import {Router, Route, useRouterHistory} from 'react-router'
import {createHistory} from 'history'
import {syncHistoryWithStore} from 'react-router-redux'

import App from 'src/App'
import AlertsApp from 'src/alerts'
import CheckSources from 'src/CheckSources'
import {HostsPage, HostPage} from 'src/hosts'
import {KubernetesPage} from 'src/kubernetes'
import {Login, UserIsAuthenticated, Authenticated, UserIsNotAuthenticated} from 'src/auth'
import {KapacitorPage, KapacitorRulePage, KapacitorRulesPage, KapacitorTasksPage} from 'src/kapacitor'
import DataExplorer from 'src/data_explorer'
import {DashboardsPage, DashboardPage} from 'src/dashboards'
import {CreateSource, SourcePage, ManageSources} from 'src/sources'
import {AdminPage} from 'src/admin'
import NotFound from 'src/shared/components/NotFound'
import configureStore from 'src/store/configureStore'
import {getMe, getSources} from 'shared/apis'
import {authRequested, authReceived, meRequested, meReceived} from 'shared/actions/auth'
import {disablePresentationMode} from 'shared/actions/app'
import {publishNotification} from 'shared/actions/notifications'
import {loadLocalStorage} from './localStorage'

import 'src/style/chronograf.scss'

import {HTTP_FORBIDDEN, HEARTBEAT_INTERVAL} from 'shared/constants'

const rootNode = document.getElementById('react-root')

let browserHistory
const basepath = rootNode.dataset.basepath
window.basepath = basepath
if (basepath) {
  browserHistory = useRouterHistory(createHistory)({
    basename: basepath, // this is written in when available by the URL prefixer middleware
  })
} else {
  browserHistory = useRouterHistory(createHistory)({
    basename: "",
  })
}

const store = configureStore(loadLocalStorage(), browserHistory)
const {dispatch} = store

browserHistory.listen(() => {
  dispatch(disablePresentationMode())
})

window.addEventListener('keyup', (event) => {
  if (event.key === 'Escape') {
    dispatch(disablePresentationMode())
  }
})

const history = syncHistoryWithStore(browserHistory, store)

const Root = React.createClass({
  getInitialState() {
    return {
      hasDeterminedAuth: false,
    }
  },

  componentWillMount() {
    this.checkAuth()
  },

  activeSource(sources) {
    const defaultSource = sources.find((s) => s.default)
    if (defaultSource && defaultSource.id) {
      return defaultSource
    }
    return sources[0]
  },

  async checkAuth() {
    dispatch(authRequested())
    dispatch(meRequested())
    try {
      await this.startHeartbeat({shouldDispatchResponse: true})
    } catch (error) {
      console.error(error)
    }
  },

  // TODO: use once auth router is working
  redirectFromRoot(_, replace, callback) {
    getSources().then(({data: {sources}}) => {
      if (sources && sources.length) {
        const path = `/sources/${this.activeSource(sources).id}/hosts`
        replace(path)
      }
      callback()
    })
  },

  async startHeartbeat({shouldDispatchResponse}) {
    try {
      const {data: me, auth} = await getMe()
      if (shouldDispatchResponse) {
        dispatch(authReceived(auth))
        dispatch(meReceived(me))
      }

      setTimeout(this.startHeartbeat.bind(null, {shouldDispatchResponse: false}), HEARTBEAT_INTERVAL)
    } catch (error) {
      if (error.auth) {
        dispatch(authReceived(error.auth))
        dispatch(meReceived(null))
      }
      if (error.status === HTTP_FORBIDDEN) {
        dispatch(publishNotification('error', 'Session timed out. Please login again.'))
      } else {
        dispatch(publishNotification('error', 'Cannot communicate with server.'))
      }
    }
  },

  render() {
    return (
      <Provider store={store}>
        <Router history={history}>
          <Route path="/" component={UserIsAuthenticated(CreateSource)/* add back onEnter={this.redirectFromRoot} */} />
          <Route path="login" component={UserIsNotAuthenticated(Login)} />
          <Route path="sources/new" component={UserIsAuthenticated(CreateSource)} />
          <Route path="sources/:sourceID" component={UserIsAuthenticated(App)}>
            <Route component={UserIsAuthenticated(CheckSources)}>
              <Route component={Authenticated}>
                <Route path="manage-sources" component={ManageSources} />
                <Route path="manage-sources/new" component={SourcePage} />
                <Route path="manage-sources/:id/edit" component={SourcePage} />
                <Route path="chronograf/data-explorer" component={DataExplorer} />
                <Route path="hosts" component={HostsPage} />
                <Route path="hosts/:hostID" component={HostPage} />
                <Route path="kubernetes" component={KubernetesPage} />
                <Route path="kapacitor-config" component={KapacitorPage} />
                <Route path="kapacitor-tasks" component={KapacitorTasksPage} />
                <Route path="alerts" component={AlertsApp} />
                <Route path="dashboards" component={DashboardsPage} />
                <Route path="dashboards/:dashboardID" component={DashboardPage} />
                <Route path="alert-rules" component={KapacitorRulesPage} />
                <Route path="alert-rules/:ruleID" component={KapacitorRulePage} />
                <Route path="alert-rules/new" component={KapacitorRulePage} />
                <Route path="admin" component={AdminPage} />
              </Route>
            </Route>
          </Route>
          <Route path="*" component={NotFound} />
        </Router>
      </Provider>
    )
  },
})

if (rootNode) {
  render(<Root />, rootNode)
}
