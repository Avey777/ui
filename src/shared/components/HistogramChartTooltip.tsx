import React, {SFC, CSSProperties} from 'react'

import {HoverData, ColorScale} from 'src/types/histogram'

interface Props {
  data: HoverData
  colorScale: ColorScale
}

const HistogramChartTooltip: SFC<Props> = props => {
  const {colorScale} = props
  const {data, x, y, anchor = 'left'} = props.data

  const tooltipStyle: CSSProperties = {
    position: 'fixed',
    top: y,
  }

  if (anchor === 'left') {
    tooltipStyle.left = x
  } else {
    tooltipStyle.right = x
  }

  return (
    <div className="histogram-chart-tooltip" style={tooltipStyle}>
      <div className="histogram-chart-tooltip--column">
        {data.map(d => (
          <div key={d.key} style={{color: colorScale(d.group)}}>
            {d.value}
          </div>
        ))}
      </div>
      <div className="histogram-chart-tooltip--column">
        {data.map(d => (
          <div key={d.key} style={{color: colorScale(d.group)}}>
            {d.group}
          </div>
        ))}
      </div>
    </div>
  )
}

export default HistogramChartTooltip
