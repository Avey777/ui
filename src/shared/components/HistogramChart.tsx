import React, {PureComponent, MouseEvent} from 'react'
import _ from 'lodash'
import {scaleLinear, scaleTime, ScaleLinear, ScaleTime} from 'd3-scale'

import HistogramChartAxes from 'src/shared/components/HistogramChartAxes'
import HistogramChartBars from 'src/shared/components/HistogramChartBars'
import HistogramChartTooltip from 'src/shared/components/HistogramChartTooltip'
import HistogramChartSkeleton from 'src/shared/components/HistogramChartSkeleton'
import XBrush from 'src/shared/components/XBrush'

import extentBy from 'src/utils/extentBy'
import {getDeep} from 'src/utils/wrappers'

import {RemoteDataState} from 'src/types'
import {
  TimePeriod,
  HistogramData,
  HistogramDatum,
  Margins,
  TooltipAnchor,
} from 'src/types/histogram'

const PADDING_TOP = 0.2
const TOOLTIP_HORIZONTAL_MARGIN = 5
const TOOLTIP_REFLECT_DIST = 100

// Rather than use these magical constants, we could also render a digit and
// capture its measured width with as state before rendering anything else.
// Doing so would be robust but overkill.
const DIGIT_WIDTH = 7
const PERIOD_DIGIT_WIDTH = 4

interface Props {
  data: HistogramData
  dataStatus: RemoteDataState
  width: number
  height: number
  onZoom: (TimePeriod) => void
}

interface State {
  hoverX: number
  hoverY: number
  hoverDatum?: HistogramDatum
  hoverAnchor: TooltipAnchor
}

class HistogramChart extends PureComponent<Props, State> {
  constructor(props) {
    super(props)

    this.state = {hoverX: -1, hoverY: -1, hoverAnchor: 'left'}
  }

  public render() {
    const {width, height, data} = this.props
    const {margins} = this

    if (width === 0 || height === 0) {
      return null
    }

    if (!data.length) {
      return (
        <HistogramChartSkeleton
          width={width}
          height={height}
          margins={margins}
        />
      )
    }

    const {hoverDatum, hoverX, hoverY, hoverAnchor} = this.state
    const {
      xScale,
      yScale,
      adjustedWidth,
      adjustedHeight,
      bodyTransform,
      loadingClass,
    } = this

    return (
      <>
        <svg
          width={width}
          height={height}
          className={`histogram-chart ${loadingClass}`}
          onMouseOver={this.handleMouseMove}
          onMouseOut={this.handleMouseOut}
        >
          <defs>
            <clipPath id="histogram-chart--bars-clip">
              <rect x="0" y="0" width={adjustedWidth} height={adjustedHeight} />
            </clipPath>
          </defs>
          <g className="histogram-chart--axes">
            <HistogramChartAxes
              width={width}
              height={height}
              margins={margins}
              xScale={xScale}
              yScale={yScale}
            />
          </g>
          <g className="histogram-chart--brush" transform={bodyTransform}>
            <XBrush
              xScale={xScale}
              width={adjustedWidth}
              height={adjustedHeight}
              onBrush={this.handleBrush}
            />
          </g>
          <g
            transform={bodyTransform}
            className="histogram-chart--bars"
            clipPath="url(#histogram-chart--bars-clip)"
          >
            <HistogramChartBars
              width={adjustedWidth}
              height={adjustedHeight}
              data={data}
              xScale={xScale}
              yScale={yScale}
            />
          </g>
        </svg>
        <HistogramChartTooltip
          datum={hoverDatum}
          x={hoverX}
          y={hoverY}
          anchor={hoverAnchor}
        />
      </>
    )
  }

  private get xScale(): ScaleTime<number, number> {
    const {adjustedWidth} = this
    const {data} = this.props

    const [t0, t1] = extentBy(data, d => d.time)

    return scaleTime()
      .domain([new Date(t0.time), new Date(t1.time)])
      .range([0, adjustedWidth])
  }

  private get yScale(): ScaleLinear<number, number> {
    const {adjustedHeight, maxAggregateCount} = this

    return scaleLinear()
      .domain([0, maxAggregateCount + PADDING_TOP * maxAggregateCount])
      .range([adjustedHeight, 0])
  }

  private get adjustedWidth(): number {
    const {margins} = this

    return this.props.width - margins.left - margins.right
  }

  private get adjustedHeight(): number {
    const {margins} = this

    return this.props.height - margins.top - margins.bottom
  }

  private get bodyTransform(): string {
    const {margins} = this

    return `translate(${margins.left}, ${margins.top})`
  }

  private get margins(): Margins {
    const {maxAggregateCount} = this

    const domainTop = maxAggregateCount + PADDING_TOP * maxAggregateCount
    const left = domainTop.toString().length * DIGIT_WIDTH + PERIOD_DIGIT_WIDTH

    return {top: 5, right: 0, bottom: 20, left}
  }

  private get maxAggregateCount(): number {
    const {data} = this.props

    if (!data.length) {
      return 0
    }

    const groups = _.groupBy(data, 'time')
    const counts = Object.values(groups).map(group =>
      group.reduce((sum, current) => sum + current.value, 0)
    )

    return Math.max(...counts)
  }

  private get loadingClass(): string {
    const {dataStatus} = this.props

    return dataStatus === RemoteDataState.Loading ? 'loading' : ''
  }

  private handleBrush = (t: TimePeriod): void => {
    this.props.onZoom(t)
    this.setState({hoverDatum: null})
  }

  private handleMouseMove = (e: MouseEvent<SVGElement>): void => {
    const key = getDeep<string>(e, 'target.dataset.key', '')

    if (!key) {
      return
    }

    const {data} = this.props
    const hoverDatum = data.find(d => d.key === key)

    if (!hoverDatum) {
      return
    }

    const bar = e.target as SVGRectElement
    const barRect = bar.getBoundingClientRect()
    const barRectHeight = barRect.bottom - barRect.top
    const hoverY = barRect.top + barRectHeight / 2

    let hoverX = barRect.right + TOOLTIP_HORIZONTAL_MARGIN
    let hoverAnchor: TooltipAnchor = 'left'

    if (hoverX >= window.innerWidth - TOOLTIP_REFLECT_DIST) {
      hoverX = window.innerWidth - barRect.left + TOOLTIP_HORIZONTAL_MARGIN
      hoverAnchor = 'right'
    }

    this.setState({hoverDatum, hoverX, hoverY, hoverAnchor})
  }

  private handleMouseOut = (): void => {
    this.setState({hoverDatum: null})
  }
}

export default HistogramChart
