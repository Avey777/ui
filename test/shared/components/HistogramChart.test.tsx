import React from 'react'
import {mount, shallow} from 'enzyme'

import HistogramChart from 'src/shared/components/HistogramChart'
import HistogramChartTooltip from 'src/shared/components/HistogramChartTooltip'

import {RemoteDataState} from 'src/types'

describe('HistogramChart', () => {
  test('displays a HistogramChartSkeleton if empty data is passed', () => {
    const props = {
      data: [],
      dataStatus: RemoteDataState.Done,
      width: 600,
      height: 400,
      colorScale: () => 'blue',
      onZoom: () => {},
    }

    const wrapper = mount(<HistogramChart {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  test('displays a nothing if passed width and height of 0', () => {
    const props = {
      data: [],
      dataStatus: RemoteDataState.Done,
      width: 0,
      height: 0,
      colorScale: () => 'blue',
      onZoom: () => {},
    }

    const wrapper = mount(<HistogramChart {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  test('displays the visualization with bars if nonempty data is passed', () => {
    const props = {
      data: [
        {key: '0', time: 0, value: 0, group: 'a'},
        {key: '1', time: 1, value: 1, group: 'a'},
        {key: '2', time: 2, value: 2, group: 'b'},
      ],
      dataStatus: RemoteDataState.Done,
      width: 600,
      height: 400,
      colorScale: () => 'blue',
      onZoom: () => {},
    }

    const wrapper = mount(<HistogramChart {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  test('displays a HistogramChartTooltip when hovering over bars', () => {
    const props = {
      data: [
        {key: '0', time: 0, value: 0, group: 'a'},
        {key: '1', time: 1, value: 1, group: 'a'},
        {key: '2', time: 2, value: 2, group: 'b'},
      ],
      dataStatus: RemoteDataState.Done,
      width: 600,
      height: 400,
      colorScale: () => 'blue',
      onZoom: () => {},
    }

    const wrapper = mount(<HistogramChart {...props} />)

    const fakeMouseOverEvent = {
      target: {
        getBoundingClientRect() {
          return {top: 10, right: 10, bottom: 5, left: 5}
        },
      },
    }

    wrapper
      .find('.histogram-chart-bars--bars')
      .first()
      .simulate('mouseover', fakeMouseOverEvent)

    const tooltip = wrapper.find(HistogramChartTooltip)

    expect(tooltip).toMatchSnapshot()
  })

  test('has a "loading" class if data is reloading', () => {
    const props = {
      data: [{key: '', time: 0, value: 0, group: ''}],
      dataStatus: RemoteDataState.Loading,
      width: 600,
      height: 400,
      colorScale: () => 'blue',
      onZoom: () => {},
    }

    const wrapper = shallow(<HistogramChart {...props} />)

    expect(wrapper.find('.histogram-chart').hasClass('loading')).toBe(true)
  })
})
