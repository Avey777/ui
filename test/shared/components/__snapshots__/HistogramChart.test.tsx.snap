// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistogramChart displays a HistogramChartSkeleton if empty data is passed 1`] = `
<HistogramChart
  data={Array []}
  dataStatus="Done"
  height={400}
  onZoom={[Function]}
  width={600}
>
  <HistogramChartSkeleton
    height={400}
    margins={
      Object {
        "bottom": 20,
        "left": 11,
        "right": 0,
        "top": 5,
      }
    }
    width={600}
  >
    <svg
      className="histogram-chart-skeleton"
      height={400}
      width={600}
    >
      <line
        className="y-tick"
        key="0"
        x1={11}
        x2={600}
        y1={380}
        y2={380}
      />
      <line
        className="y-tick"
        key="1"
        x1={11}
        x2={600}
        y1={305}
        y2={305}
      />
      <line
        className="y-tick"
        key="2"
        x1={11}
        x2={600}
        y1={230}
        y2={230}
      />
      <line
        className="y-tick"
        key="3"
        x1={11}
        x2={600}
        y1={155}
        y2={155}
      />
      <line
        className="y-tick"
        key="4"
        x1={11}
        x2={600}
        y1={80}
        y2={80}
      />
    </svg>
  </HistogramChartSkeleton>
</HistogramChart>
`;

exports[`HistogramChart displays a HistogramChartTooltip when hovering over bars 1`] = `
<HistogramChartTooltip
  anchor="left"
  datum={
    Object {
      "group": "a",
      "key": "0",
      "time": 0,
      "value": 0,
    }
  }
  x={15}
  y={7.5}
>
  <div
    className="histogram-chart-tooltip"
    data-group="a"
    style={
      Object {
        "left": 15,
        "position": "fixed",
        "top": 7.5,
      }
    }
  >
    <div
      className="histogram-chart-tooltip--value"
    >
      0
    </div>
    <div
      className="histogram-chart-tooltip--group"
    >
      a
    </div>
  </div>
</HistogramChartTooltip>
`;

exports[`HistogramChart displays a nothing if passed width and height of 0 1`] = `
<HistogramChart
  data={Array []}
  dataStatus="Done"
  height={0}
  onZoom={[Function]}
  width={0}
/>
`;

exports[`HistogramChart displays the visualization with bars if nonempty data is passed 1`] = `
<HistogramChart
  data={
    Array [
      Object {
        "group": "a",
        "key": "0",
        "time": 0,
        "value": 0,
      },
      Object {
        "group": "a",
        "key": "1",
        "time": 1,
        "value": 1,
      },
      Object {
        "group": "b",
        "key": "2",
        "time": 2,
        "value": 2,
      },
    ]
  }
  dataStatus="Done"
  height={400}
  onZoom={[Function]}
  width={600}
>
  <svg
    className="histogram-chart "
    height={400}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    width={600}
  >
    <defs>
      <clipPath
        id="histogram-chart--bars-clip"
      >
        <rect
          height={375}
          width={575}
          x="0"
          y="0"
        />
      </clipPath>
    </defs>
    <g
      className="histogram-chart--axes"
    >
      <HistogramChartAxes
        height={400}
        margins={
          Object {
            "bottom": 20,
            "left": 25,
            "right": 0,
            "top": 5,
          }
        }
        width={600}
        xScale={[Function]}
        yScale={[Function]}
      >
        <line
          className="y-tick"
          key="0-25-625-380"
          x1={25}
          x2={625}
          y1={380}
          y2={380}
        />
        <line
          className="y-tick"
          key="0.5-25-625-301.875"
          x1={25}
          x2={625}
          y1={301.875}
          y2={301.875}
        />
        <line
          className="y-tick"
          key="1-25-625-223.75"
          x1={25}
          x2={625}
          y1={223.75}
          y2={223.75}
        />
        <line
          className="y-tick"
          key="1.5-25-625-145.625"
          x1={25}
          x2={625}
          y1={145.625}
          y2={145.625}
        />
        <line
          className="y-tick"
          key="2-25-625-67.5"
          x1={25}
          x2={625}
          y1={67.5}
          y2={67.5}
        />
        <text
          className="y-label"
          key="0-25-625-380"
          x={20}
          y={380}
        >
          0
        </text>
        <text
          className="y-label"
          key="0.5-25-625-301.875"
          x={20}
          y={301.875}
        >
          0.5
        </text>
        <text
          className="y-label"
          key="1-25-625-223.75"
          x={20}
          y={223.75}
        >
          1
        </text>
        <text
          className="y-label"
          key="1.5-25-625-145.625"
          x={20}
          y={145.625}
        >
          1.5
        </text>
        <text
          className="y-label"
          key="2-25-625-67.5"
          x={20}
          y={67.5}
        >
          2
        </text>
        <text
          className="x-label"
          key=".001-287.5-388"
          x={287.5}
          y={388}
        >
          .001
        </text>
        <text
          className="x-label"
          key=".002-575-388"
          x={575}
          y={388}
        >
          .002
        </text>
      </HistogramChartAxes>
    </g>
    <g
      className="histogram-chart--brush"
      transform="translate(25, 5)"
    >
      <XBrush
        height={375}
        onBrush={[Function]}
        width={575}
        xScale={[Function]}
      >
        <rect
          className="x-brush--area"
          height={375}
          onDoubleClick={[Function]}
          onMouseDown={[Function]}
          width={575}
        />
      </XBrush>
    </g>
    <g
      className="histogram-chart--bars"
      clipPath="url(#histogram-chart--bars-clip)"
      transform="translate(25, 5)"
    >
      <HistogramChartBars
        data={
          Array [
            Object {
              "group": "a",
              "key": "0",
              "time": 0,
              "value": 0,
            },
            Object {
              "group": "a",
              "key": "1",
              "time": 1,
              "value": 1,
            },
            Object {
              "group": "b",
              "key": "2",
              "time": 2,
              "value": 2,
            },
          ]
        }
        height={375}
        width={575}
        xScale={[Function]}
        yScale={[Function]}
      >
        <g
          className="histogram-chart-bars--bars"
          key="1-1-193.5"
        >
          <defs>
            <clipPath
              id="histogram-chart-bars--clip-1-1-193.5"
            >
              <rect
                height={160.25}
                rx={4}
                ry={4}
                width={188}
                x={193.5}
                y={218.75}
              />
            </clipPath>
          </defs>
          <rect
            className="histogram-chart-bars--bar"
            clipPath="url(#histogram-chart-bars--clip-1-1-193.5)"
            data-group="a"
            data-key="1"
            height={156.25}
            key="1"
            width={188}
            x={193.5}
            y={218.75}
          />
        </g>
        <g
          className="histogram-chart-bars--bars"
          key="2-2-481"
        >
          <defs>
            <clipPath
              id="histogram-chart-bars--clip-2-2-481"
            >
              <rect
                height={316.5}
                rx={4}
                ry={4}
                width={188}
                x={481}
                y={62.5}
              />
            </clipPath>
          </defs>
          <rect
            className="histogram-chart-bars--bar"
            clipPath="url(#histogram-chart-bars--clip-2-2-481)"
            data-group="b"
            data-key="2"
            height={312.5}
            key="2"
            width={188}
            x={481}
            y={62.5}
          />
        </g>
      </HistogramChartBars>
    </g>
  </svg>
  <HistogramChartTooltip
    anchor="left"
    x={-1}
    y={-1}
  />
</HistogramChart>
`;
